### PR TITLE
feat: apis for netifyd

### PR DESCRIPTION
**Pull request by Tbaile**
Added dpi methods
- index application/protocol
- index rules
- add rule
- delete rule
- edit rule

### CODE REVIEW

File: src/nethsec/dpi/__init__.py [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+    """
+    result = __load_blocklist()
+
+    if search is not None:
+        # lower string so we can do a case-insensitive search
+        search = search.lower()
+        # I'm aware it's far from a readable code, but list comprehension is the fastest way to filter.
+        result = [item for item in result if
+                  item.get('name', '').lower().startswith(search) or
+                  item.get('category', {}).get('name', '').lower().startswith(search)]
```
Comment (gsanchietti) [inline src/nethsec/dpi/__init__.py:167]: search also inside the string, not only at the beginning

File: src/nethsec/dpi/__init__.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+
+    return rules
+
+
+def __generate_rule_name(e_uci):
+    counter = 0
+    for role in utils.get_all_by_type(e_uci, 'dpi', 'rule').keys():
+        if f'rule{counter}' == role:
+            counter += 1
+    return f'rule{counter}'
```
Comment (gsanchietti) [inline src/nethsec/dpi/__init__.py:234]: Why just not using  a random id from utils package?

File: src/nethsec/dpi/__init__.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+
+def __save_rule_data(e_uci: EUci, config_name: str, enabled: bool, interface: str, applications: list[str],
+                     protocols: list[str]):
+    e_uci.set('dpi', config_name, 'enabled', enabled)
+    e_uci.set('dpi', config_name, 'interface', interface)
+    e_uci.set('dpi', config_name, 'application', [f'netify.{application}' for application in applications])
+    e_uci.set('dpi', config_name, 'protocol', protocols)
+
+
+def add_rule(e_uci: EUci, enabled: bool, interface: str, applications: list[str], protocols: list[str]) -> str:
```
Comment (gsanchietti) [inline src/nethsec/dpi/__init__.py:245]: Rules can have 4 types of actions: https://github.com/NethServer/nethsecurity/blob/fc08343740793edba25562fd37a3ef962fdfe5f0/packages/ns-dpi/files/dpi-config#L75

The add_rule method must take the action as parameter.

File: src/nethsec/dpi/__init__.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+
+    Args:
+      - e_uci: euci instance
+      - config_name: config name of the rule to delete
+    """
+    e_uci.delete('dpi', config_name)
+    e_uci.save('dpi')
+
+
+def edit_rule(e_uci: EUci, config_name: str, enabled: bool, interface: str, applications: list[str],
```
Comment (gsanchietti) [inline src/nethsec/dpi/__init__.py:280]: Please provide `action` parameter.

File: src/nethsec/dpi/__init__.py [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+            blocklist = __load_blocklist()
+            # get content of rule
+            rule = fetch_rules[rule_name]
+            # prepare the data to append to rules
+            data_rule = dict[str]()
+            data_rule['config-name'] = rule_name
+            data_rule['enabled'] = rule.get('enabled', '1') == '1'
+            data_rule['interface'] = rule.get('interface', '*')
+            # get the blocked applications/protocols
+            data_rule['blocks'] = list[dict[str]]()
```
Comment (gsanchietti) [inline src/nethsec/dpi/__init__.py:204]: Use a more generic field name like `target` and add and `action` field.

File: src/nethsec/dpi/__init__.py [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+      - applications: array of applications to block
+      - protocols: array of protocols to block
+
+    Raises
+        - ValidationError: if the config name is invalid
+    """
+    if e_uci.get('dpi', config_name, default=None) is None:
+        raise ValidationError('config-name', 'invalid', config_name)
+
+    __save_rule_data(e_uci, config_name, enabled, interface, applications, protocols)
```
Comment (gsanchietti) [inline src/nethsec/dpi/__init__.py:300]: Add `action` parameter

File: src/nethsec/dpi/__init__.py [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+      - interface: interface to listen and apply the rule on
+      - applications: list of applications to block
+      - protocols: list of protocols to block
+
+    Returns:
+        config name of the rule created
+    """
+    rule_name = __generate_rule_name(e_uci)
+    e_uci.set('dpi', rule_name, 'rule')
+    e_uci.set('dpi', rule_name, 'action', 'block')
```
Comment (gsanchietti) [inline src/nethsec/dpi/__init__.py:262]: Do not set the action only to block

File: src/nethsec/dpi/__init__.py [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+        if f'rule{counter}' == role:
+            counter += 1
+    return f'rule{counter}'
+
+
+def __save_rule_data(e_uci: EUci, config_name: str, enabled: bool, interface: str, applications: list[str],
+                     protocols: list[str]):
+    e_uci.set('dpi', config_name, 'enabled', enabled)
+    e_uci.set('dpi', config_name, 'interface', interface)
+    e_uci.set('dpi', config_name, 'application', [f'netify.{application}' for application in applications])
```
Comment (gsanchietti) [inline src/nethsec/dpi/__init__.py:241]: Please preserve the original name, the `netify` prefix could change. Eg.: if we create our own signatures.

File: src/nethsec/dpi/__init__.py [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+            data_rule['enabled'] = rule.get('enabled', '1') == '1'
+            data_rule['interface'] = rule.get('interface', '*')
+            # get the blocked applications/protocols
+            data_rule['blocks'] = list[dict[str]]()
+
+            # filter by application
+            application_blocklist = [item for item in blocklist if item['type'] == 'application']
+            for application in rule.get('application', []):
+                found_app = [item for item in application_blocklist if
+                             item['name'] == application.removeprefix('netify.')]
```
Comment (gsanchietti) [inline src/nethsec/dpi/__init__.py:210]: Do not use 'netify.' as a fixed prefix.

**Review by gsanchietti — CHANGES_REQUESTED** [review]
Main changes:
- support other actions other than block: we are writing a library, so it should expose all functions. The APIs for now can implement only the block action
- always use the full application name inside the rules, if a pretty name is needed by the APIs, just return an additional field

File: src/nethsec/dpi/__init__.py (hunk 8 above)
Comment (Tbaile) [reply src/nethsec/dpi/__init__.py:241]: Will return the full identifier so it's future-proof

File: src/nethsec/dpi/__init__.py (hunk 9 above)
Comment (Tbaile) [reply src/nethsec/dpi/__init__.py:210]: Same as above, returning full identifier of application rule, letting the UI handle this split

**Review by gsanchietti — APPROVED** [review]

**Pull request merged**